lib/ddtrace/sampler.rb

```diff
@@ -194,11 +204,24 @@ def sample!(span)
       if span.sampled
         # If priority sampling has already been applied upstream, use that, otherwise...
         unless priority_assigned_upstream?(span)
+          pre_sample_rate_metric = span.get_metric(SAMPLE_RATE_METRIC_KEY)
+
           # Roll the dice and determine whether how we set the priority.
+          priority = priority_sample!(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+
           # NOTE: We'll want to leave `span.sampled = true` here; all spans for priority sampling must
           #       be sent to the agent. Otherwise metrics for traces will not be accurate, since the
           #       agent will have an incomplete dataset.
-          priority = priority_sample(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+          #
+          #       We also ensure that the agent knows we that our `post_sampler` is not performing true sampling,
+          #       to avoid erroneous metric upscaling.
+          span.sampled = true
+          if pre_sample_rate_metric
+            span.set_metric(SAMPLE_RATE_METRIC_KEY, pre_sample_rate_metric) # Restore true sampling metric
+          else
+            span.clear_metric(SAMPLE_RATE_METRIC_KEY)
+          end
```

Comment:
This class is growing in complexity in order to support the `RuleSampler` as a `post_sampler`.
It seems nice to keep all priority sampling concerns here, and not have any other sampler worry about it, but I can see an argument for moving away from having the current `PrioritySampler` try to do it all.

I'm open to suggestions for different approaches to simplify this approach.

The current implementation in this PR works, but trying to maintain compatibility with all existing samplers, alongside the new rule sampler, has created an arms-race for the `PrioritySampler` class. I believe this implementation will be much simpler if we move away from having an agent one day.

Reply:
In the interest of time, we might want to keep what we have here if its working, and refactor in the future. I would just make sure you have sufficient tests to verify all these edge cases and conditions introduced by this complexity such that when we do refactor later, we won't introduce any bugs to the sampling logic. If we can do that, then I think its okay to let this be.